aten/src/ATen/native/cpu/utils.h

```diff
@@ -53,6 +53,31 @@ inline bool data_index_step(T& x, const T& X, Args&&... args) {
   return false;
 }
 
+// Helper struct for compile time loop unrolling
+template <int i>
+struct compile_time_for {
+  template <typename Lambda, typename... Args>
+  inline static void op(const Lambda& function, Args... args) {
+    compile_time_for<i - 1>::op(function, args...);
+    function(std::integral_constant<int, i - 1>{}, args...);
+  }
+};
+
+template <>
+struct compile_time_for<1> {
+  template <typename Lambda, typename... Args>
+  inline static void op(const Lambda& function, Args... args) {
+    function(std::integral_constant<int, 0>{}, args...);
+  }
+};
+
+template <>
+struct compile_time_for<0> {
+  // 0 loops, do nothing
+  template <typename Lambda, typename... Args>
+  inline static void op(const Lambda& function, Args... args) {}
+};
+
```

Comment:
Isn't it identical to https://github.com/pytorch/pytorch/blob/9b2c35b4fe933992a1d7c9ea1765b25f2f64e35f/c10/util/Unroll.h#L13-L19

Reply:
Ohhhh! did not realize `ForcedUnroll`, many thanks !

Reply:
Not sure if it's doing the same, but if it does can be reused (or updated to C++17 goodness)

Reply:
it does the same thing! But i need to update `c10::ForcedUnroll` a little bit here.  So as to add `Args...`. Also have to make `i` as compile time constant, so the compiler can do `constexpr  int row = i / COLS;` and `constexpr  int col = i % COLS;`  for us. Integer div is very slow on CPU.